### PR TITLE
Do not raise S3BucketUnavailable when exporting and AWS credentials are missing

### PR DESCRIPTION
**Pull request by fmhoeger**
### Changed
- Improve handling of export when no AWS credentials are registered

### CODE REVIEW

**Comment by fmhoeger** [comment]
@pmcharrison We do not raise an  `S3BucketUnavailable` error when AWS credentials are missing and stop execution. Instead we only print a warning to the console. Does this match with the idea behind this issue?

If the bucket is `None` it is clearly from not being able to create or get a bucket. Or do we want to explicitly test if authentication can be done successful?

**Comment by pmcharrison** [comment]
I think we should change the logic in 

```py
def user_s3_bucket(canonical_user_id=None):
    """Get the user's S3 bucket."""
    s3 = _s3_resource()
    if not canonical_user_id:
        try:
            canonical_user_id = _get_canonical_aws_user_id(s3)
        except botocore.exceptions.ClientError:
            return None
```

and remove this part where we catch ClientError and return None. I think it's a bad idea to silently fail in such cases and it's not intuitive that None should be interpreted as an authentication error.

> @pmcharrison We do not raise an S3BucketUnavailable error when AWS credentials are missing and stop execution. Instead we only print a warning to the console. Does this match with the idea behind this issue?

I don't know what your words mean, sorry.... when you say "we do not raise", are you talking about existing behaviour, or about proposed behaviour?

In any case, my proposal is that if the AWS credentials are missing we do not stop execution, we just print a warning.

**Comment by pmcharrison** [comment]
i.e. 

replace

```py
def user_s3_bucket(canonical_user_id=None):
    """Get the user's S3 bucket."""
    s3 = _s3_resource()
    if not canonical_user_id:
        try:
            canonical_user_id = _get_canonical_aws_user_id(s3)
        except botocore.exceptions.ClientError:
            return None
```

with 

```py
def user_s3_bucket(canonical_user_id=None):
    """Get the user's S3 bucket."""
    s3 = _s3_resource()
    if not canonical_user_id:
        canonical_user_id = _get_canonical_aws_user_id(s3)
```

**Comment by pmcharrison** [comment]
Then bucket will never be None, so you can delete the handling logic for bucket is None.

**Comment by fmhoeger** [comment]
> Then bucket will never be None, so you can delete the handling logic for bucket is None.

But now we have to handle every every call of `user_s3_bucket()` separately to catch a possible `botocore.exceptions.ClientError`.

**Comment by fmhoeger** [comment]
OK, I added some logic to `find_experiment_export` for handling missing credentials. There are two more places where `user_s3_bucket` is called, though (`data.backup` and `command_line.__init__.awaken`).

File: dallinger/data.py [hunk 1]
```diff
@@ -281,12 +290,23 @@ def export_db_uri(id, db_uri, local, scrub_pii):
                 f" - bucket name: {bucket.name}\n"
                 f" - S3 console URL: {s3_console_url}"
             )
-        except AttributeError:
-            raise S3BucketUnavailable("Could not find an S3 bucket!")
+        else:
+            print("✘ Could not find an S3 bucket!")
```
Comment (pmcharrison) [inline dallinger/data.py:294]: It looks like this error message will be printed if no AWS credentials are found. This doesn't seem like a sensible message for that situation.

File: dallinger/data.py [hunk 2]
```diff
@@ -265,9 +270,13 @@ def export_db_uri(id, db_uri, local, scrub_pii):
...
     if not local:
-        bucket = user_s3_bucket()
         config = get_config()
-        try:
+
+        bucket = None
+        if aws_access_keys_present(config):
+            bucket = user_s3_bucket()
+
+        if bucket is not None:
```
Comment (pmcharrison) [inline dallinger/data.py:279]: Why are you testing if the bucket is None? None is no longer an output from `user_s3_bucket`. If you are testing rather whether `aws_access_keys_present` returns `True`, you should make that more explicit.

File: dallinger/data.py [hunk 3]
```diff
@@ -281,12 +290,23 @@ def export_db_uri(id, db_uri, local, scrub_pii):
...
             )
-        except AttributeError:
-            raise S3BucketUnavailable("Could not find an S3 bucket!")
+        else:
+            print("✘ Could not find an S3 bucket!")
 
     return path_to_data
 
 
+def aws_access_keys_present(config):
```
Comment (pmcharrison) [inline dallinger/data.py:299]: This function should return a boolean, I think. I don't think it should print a warning message like it currently does. You want to write instead something like 

```py
if not aws_access_keys_present(config):
    logger.warn("AWS credentials not present, skipping export to S3.")
```

File: dallinger/data.py [hunk 4]
```diff
@@ -84,17 +84,22 @@ def find_experiment_export(app_id):
...
-    for bucket in buckets:
-        if bucket is None:
-            continue
-        try:
-            bucket.download_file(data_filename, path_to_data)
-        except botocore.exceptions.ClientError:
-            pass
-        else:
-            return path_to_data
+    buckets = []
```
Comment (pmcharrison) [inline dallinger/data.py:90]: You could minimise the diff by instead just adding 

```py
if not aws_access_keys_present(config):
    return
```

File: dallinger/data.py (hunk 1 above)
Comment (fmhoeger) [reply dallinger/data.py]: OK, I think I finally got what solution you were looking for :+1:

File: dallinger/data.py (hunk 2 above)
Comment (fmhoeger) [reply dallinger/data.py]: Addressed this one, too.

File: dallinger/data.py (hunk 3 above)
Comment (fmhoeger) [reply dallinger/data.py:302]: Done.

Comment (fmhoeger) [reply dallinger/data.py:302]: Unfortunately, no matter which logging level I use, nothing gets printed on the console. So I reverted to `print`.

**Review by pmcharrison — COMMENTED** [review]
LGTM!

**Review by pmcharrison — APPROVED** [review]

**Pull request merged**